lib/websocket.c

```diff
@@ -222,6 +224,12 @@ void h2o_websocket_proceed(h2o_websocket_conn_t *conn)
         }
     } while (handled);
 
+    if (conn->_write_buf.cnt > 0) {
+        /* write */
+        h2o_socket_write(conn->sock, conn->_write_buf.bufs, conn->_write_buf.cnt, on_write_complete);
```

Comment:
You cannot write to a socket that has a pending-write. In the loop above, there is a check for that, but nothing here.

I would suggest instead to move the code that writes the pending write to the beginning of `on_write_complete`. Please also make sure that the buffers are not freed until `on_write_complete` is called.

Reply:
when the "pending-write" can happen?
this is the only place where `h2o_socket_write` called.
if pending-write exist, `conn->_write_buf.cnt > 0` can be true?

Reply:
Thank you for elaborating. Now I understand the intent of the code. It looks fine to me.